### ISSUE TITLE
[Data] `map_batches` hard to use and debug

### ISSUE THREAD

**Issue by maxpumperla** [ray-project/ray#34007]
### Description

To properly define a map function, you have to know the type of your Dataset. In our [object detection CUJ](https://docs.google.com/document/d/1v5Kzp_o3LpgYU1ZFa7-7H3uYDyDLqaDG7epUkjSgwh8/edit#), I was loading images from S3 using

```
import ray
path = "s3://air-example-data-2/movie-image-small-filesize-1GB"
ds = ray.data.read_images(path)
ds
```

The individual rows here, using `take` are of the form `{"image": <numpy array>}`, and the batch format seems to be pandas, according to:

```
ds.default_batch_format()
```

The problem is that I need to know the pandas API to define a map function, e.g. using a feature `extractor` and `model` from HF and trying:

```
def map_fn(batch: pd.DataFrame) -> pd.DataFrame:
    images = batch["image"]
    
    inputs = extractor(images=images, return_tensors="pt")
    outputs = model(**inputs)
    target_sizes = torch.tensor([image.size[::-1]])
    return extractor.post_process_object_detection(outputs, threshold=0.9, target_sizes=target_sizes)[0]
```

Doesn't work and the error is difficult to read for beginners. Not everyone knows pandas just like that. The bigger issue is that it's not very clear how to test this function properly, as I'd have to create(!) a pandas dataframe first, somehow. This is a bit chicken-egg, as I might not know pandas. 

To summarize, I have a Datasets abstraction, which looks and feels like a dictionary on a per-item basis. But it has pandas type, and contains numpy arrays. But using “ds.take(...)”  returns a list of ArrowRows, which is even more confusing. How am I supposed to fix the above map_fn now? All I want to do is apply a model to a Dataset I already have -- this should be easier to handle. I suspect this is where _most_ users would just give up, which isn't good enough for us.

**Comment by jianoaix** [comment]
I think Dataset is targeting data scientist/ML users, who usually know dataframe/pandas; and pandas is usually considered as having good APIs to manipulate data. These are why we make it the default batch format to write UDFs.

**Comment by ericl** [comment]
@maxpumperla we recently added `take_batch`, which should unblock this in the short term. Can you try that out and suggest any further improvements along that line?

Separately, we are investigating consolidating the map()/take() and map_batches() to use a more similar format, so we don't have this weird "take returns a different format" problem, but this change will take more time.

**Comment by maxpumperla** [comment]
@jianoaix right, I agree with "pandas as standard API". the issue is that it's hard to understand why the generic map function you have to define _all of a sudden_ requires you to work with pandas. Nothing in `read_images` seems to indicate that this would be the case. Looks like a dict of numpy arrays.

@ericl yes!!! `take_batch` is immensely helpful. good step in the right direction.

**Comment by ericl** [comment]
Btw, in the new strict mode we are trying to make the default batch format numpy.

So basically, if you don't set batch format, you get numpy arrays of tensors / strings / objects, which hopefully is more pythonic of a default. One may also speculate that perhaps this numpy format is enough and we could avoid supporting pandas format, but that's probably wishful thinking.

**Comment by maxpumperla** [comment]
@ericl yeah, I think you need some sort of "named dataset" / pandas row / column / dict-like data. to me the question is: is it really worth supporting "pandas" as format on the nose? unclear how much of the pandas API people will use in practice on their batches.

**Comment by cszhu** [comment]
This P2 issue has seen no activity in the past 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please comment and remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
